WcaOnRails/spec/support/be_invalid_with_errors.rb

```diff
@@ -0,0 +1,15 @@
+# frozen_string_literal: true
+
+RSpec::Matchers.define :be_invalid_with_errors do |errors|
+  match do |ar_object|
+    !ar_object.valid? && errors.all? { |k, v| ar_object.errors[k] == v }
+  end
+
+  failure_message do |ar_object|
+    if ar_object.valid?
+      "expected to find the following errors: #{errors}"
+    else
+      "expected that #{ar_object} would be invalid"
+    end
+  end
+end
```

Comment:
Yes, I count 68 places where this pattern shows up!

<details>
<summary>That's a lot!</summary>

```
~/gitting/worldcubeassociation.org/WcaOnRails @kaladin> git grep -A 1 "expect.*be_invalid" | grep -B 1 "errors"
spec/helpers/application_helper_spec.rb:      expect(user).to be_invalid
spec/helpers/application_helper_spec.rb-      expect(user.errors.messages[:wca_id]).to eq ["not found"]
--
spec/models/competition_event_spec.rb:      expect(competition_event).to be_invalid_with_errors(rounds: ["[3] is wrong"])
--
--
spec/models/competition_event_spec.rb:      expect(competition_event).to be_invalid_with_errors(rounds: ["[1, 3] is wrong"])
--
--
spec/models/competition_event_spec.rb:      expect(competition_event).to be_invalid_with_errors(rounds: ["[1, 2, 2, 3] is wrong"])
--
--
spec/models/competition_event_spec.rb:      expect(competition_event).to be_invalid_with_errors(rounds: ["[1, 3] is wrong"])
--
spec/models/competition_spec.rb:      expect(competition).to be_invalid
spec/models/competition_spec.rb-      expect(competition.errors.messages[:delegate_ids]).to eq ["are not all delegates"]
--
spec/models/competition_spec.rb:    expect(competition).to be_invalid
spec/models/competition_spec.rb-    expect(competition.errors.messages[:registration_close]).to eq ["registration close must be after registration open"]
--
spec/models/competition_spec.rb:    expect(competition).to be_invalid
spec/models/competition_spec.rb-    expect(competition.errors.messages[:registration_open]).to eq ["required"]
--
spec/models/competition_spec.rb:    expect(competition).to be_invalid
spec/models/competition_spec.rb-    expect(competition.errors.messages[:registration_close]).to eq ["required"]
--
spec/models/competition_spec.rb:    expect(competition).to be_invalid
spec/models/competition_spec.rb-    expect(competition.errors.messages[:name]).to eq ["must end with a year and must contain only alphanumeric characters, dashes(-), ampersands(&), periods(.), colons(:), apostrophes('), and spaces( )"]
--
spec/models/competition_spec.rb:    expect(competition).to be_invalid
spec/models/competition_spec.rb-    expect(competition.errors.messages[:end_date]).to eq [I18n.t('competitions.errors.span_too_many_days', max_days: Competition::MAX_SPAN_DAYS)]
--
spec/models/competition_spec.rb:    expect(competition).to be_invalid
spec/models/competition_spec.rb-    expect(competition.errors.messages[:name]).to eq ["is too long (maximum is 50 characters)"]
--
spec/models/merge_people_spec.rb:    expect(merge_people).to be_invalid
spec/models/merge_people_spec.rb-    expect(merge_people.errors.messages[:person1_wca_id]).to include "This person has multiple subIds"
--
spec/models/merge_people_spec.rb:    expect(merge_people).to be_invalid
spec/models/merge_people_spec.rb-    expect(merge_people.errors.messages[:person2_wca_id]).to include "This person has multiple subIds"
--
spec/models/merge_people_spec.rb:    expect(merge_people).to be_invalid
spec/models/merge_people_spec.rb-    expect(merge_people.errors.messages[:person2_wca_id]).to eq ["Names don't match"]
--
spec/models/merge_people_spec.rb:    expect(merge_people).to be_invalid
spec/models/merge_people_spec.rb-    expect(merge_people.errors.messages[:person2_wca_id]).to eq ["Countries don't match"]
--
spec/models/merge_people_spec.rb:    expect(merge_people).to be_invalid
spec/models/merge_people_spec.rb-    expect(merge_people.errors.messages[:person2_wca_id]).to eq ["Genders don't match"]
--
spec/models/merge_people_spec.rb:    expect(merge_people).to be_invalid
spec/models/merge_people_spec.rb-    expect(merge_people.errors.messages[:person2_wca_id]).to eq ["Birthdays don't match"]
--
spec/models/merge_people_spec.rb:    expect(merge_people).to be_invalid
spec/models/merge_people_spec.rb-    expect(merge_people.errors.messages[:person2_wca_id]).to eq ["Must not have an account"]
--
spec/models/person_spec.rb:        expect(person).to be_invalid
spec/models/person_spec.rb-        expect(person.errors[:countryId]).to eq ["Cannot change the country to a country the person has already represented in the past."]
--
spec/models/poll_spec.rb:      expect(poll).to be_invalid
spec/models/poll_spec.rb-      expect(poll.errors.keys).to eq [:poll_options]
--
spec/models/registration_spec.rb:    expect(reg).to be_invalid
spec/models/registration_spec.rb-    expect(reg.errors.messages[:competition]).to eq ["Competition registration is closed"]
--
spec/models/registration_spec.rb:    expect(registration).to be_invalid
spec/models/registration_spec.rb-    expect(registration.errors[:registration_competition_events]).to eq ["must register for at least one event"]
--
spec/models/registration_spec.rb:    expect(registration).to be_invalid
spec/models/registration_spec.rb-    expect(registration.errors.messages[:guests]).to eq ["must be greater than or equal to 0"]
--
spec/models/result_spec.rb:      expect(result).to be_invalid
spec/models/result_spec.rb-      expect(result.errors.messages[:base]).to eq ["Skipped solves must all come at the end."]
--
spec/models/result_spec.rb:      expect(result).to be_invalid
spec/models/result_spec.rb-      expect(result.errors.messages[:base]).to eq ["Cannot skip all solves."]
--
spec/models/result_spec.rb:      expect(result).to be_invalid
spec/models/result_spec.rb-      expect(result.errors.messages[:value2]).to eq ["invalid"]
--
spec/models/result_spec.rb:      expect(result).to be_invalid
spec/models/result_spec.rb-      expect(result.errors.messages[:best]).to eq ["should be 42"]
--
spec/models/result_spec.rb:            expect(result).to be_invalid
spec/models/result_spec.rb-            expect(result.errors.messages[:average]).to eq ["should be 44"]
--
spec/models/result_spec.rb:            expect(result).to be_invalid
spec/models/result_spec.rb-            expect(result.errors.messages[:average]).to eq ["should be 0"]
--
spec/models/result_spec.rb:            expect(result).to be_invalid
spec/models/result_spec.rb-            expect(result.errors.messages[:average]).to eq ["should be 44"]
--
spec/models/result_spec.rb:            expect(result).to be_invalid
spec/models/result_spec.rb-            expect(result.errors.messages[:average]).to eq ["should be 43"]
--
spec/models/result_spec.rb:            expect(result).to be_invalid
spec/models/result_spec.rb-            expect(result.errors.messages[:average]).to eq ["should be 0"]
--
spec/models/result_spec.rb:            expect(result).to be_invalid
spec/models/result_spec.rb-            expect(result.errors.messages[:base]).to eq ["Expected at most 3 solves, but found 5."]
--
spec/models/result_spec.rb:            expect(result).to be_invalid
spec/models/result_spec.rb-            expect(result.errors.messages[:average]).to eq ["should be 43"]
--
spec/models/result_spec.rb:            expect(result).to be_invalid
spec/models/result_spec.rb-            expect(result.errors.messages[:base]).to eq ["Expected 3 solves, but found 1."]
--
spec/models/result_spec.rb:            expect(result).to be_invalid
spec/models/result_spec.rb-            expect(result.errors.messages[:base]).to eq ["Expected 3 solves, but found 5."]
--
spec/models/result_spec.rb:        expect(result).to be_invalid
spec/models/result_spec.rb-        expect(result.errors.messages[:average]).to eq ["should be 4233"]
--
spec/models/result_spec.rb:          expect(result).to be_invalid
spec/models/result_spec.rb-          expect(result.errors.messages[:base]).to eq ["Expected 1 solve, but found 2."]
--
spec/models/result_spec.rb:          expect(result).to be_invalid
spec/models/result_spec.rb-          expect(result.errors.messages[:base]).to eq ["Expected 2 solves, but found 3."]
--
spec/models/result_spec.rb:          expect(result).to be_invalid
spec/models/result_spec.rb-          expect(result.errors.messages[:base]).to eq ["Expected 3 solves, but found 2."]
--
spec/models/result_spec.rb:          expect(result).to be_invalid
spec/models/result_spec.rb-          expect(result.errors.messages[:base]).to eq ["Expected 3 solves, but found 2."]
--
spec/models/result_spec.rb:          expect(result).to be_invalid
spec/models/result_spec.rb-          expect(result.errors.messages[:base]).to eq ["Expected 5 solves, but found 2."]
--
spec/models/result_spec.rb:          expect(result).to be_invalid
spec/models/result_spec.rb-          expect(result.errors.messages[:base]).to eq ["Expected at most 2 solves, but found 3."]
--
spec/models/result_spec.rb:          expect(result).to be_invalid
spec/models/result_spec.rb-          expect(result.errors.messages[:base]).to eq ["Expected at most 3 solves, but found 4."]
--
spec/models/result_spec.rb:          expect(result).to be_invalid
spec/models/result_spec.rb-          expect(result.errors.messages[:base]).to eq ["Expected at most 3 solves, but found 4."]
--
spec/models/result_spec.rb:      expect(result).to be_invalid
spec/models/result_spec.rb-      expect(result.errors.messages[:value2]).to eq ["times over 10 minutes should be rounded"]
--
spec/models/result_spec.rb:        expect(result).to be_invalid
spec/models/result_spec.rb-        expect(result.errors.messages[:value1]).to eq ["should be less than or equal to 60 minutes"]
--
spec/models/result_spec.rb:        expect(result).to be_invalid
spec/models/result_spec.rb-        expect(result.errors.messages[:value1]).to eq ["should be less than or equal to 30 minutes"]
--
spec/models/round_spec.rb:      expect(round).to be_invalid_with_errors(format: ["'m' is not allowed for '333'"])
--
spec/models/user_spec.rb:      expect(user).to be_invalid
spec/models/user_spec.rb-      expect(user.errors.messages[:wca_id]).to eq [I18n.t('users.errors.wca_id_no_birthdate_html')]
--
spec/models/user_spec.rb:      expect(user).to be_invalid
spec/models/user_spec.rb-      expect(user.errors.messages[:wca_id]).to eq [I18n.t('users.errors.wca_id_no_gender_html')]
--
spec/models/user_spec.rb:      expect(user).to be_invalid
spec/models/user_spec.rb-      expect(user.errors.messages[:wca_id]).to eq ["must be unique"]
--
spec/models/user_spec.rb:      expect(user).to be_invalid
spec/models/user_spec.rb-      expect(user.errors.messages[:dob_verification]).to be_present
--
spec/models/user_spec.rb:      expect(user).to be_invalid
spec/models/user_spec.rb-      expect(user.errors.messages[:unconfirmed_wca_id]).to eq ['required']
--
spec/models/user_spec.rb:      expect(user).to be_invalid
spec/models/user_spec.rb-      expect(user.errors.messages[:unconfirmed_wca_id]).to eq ['required']
--
spec/models/user_spec.rb:      expect(user).to be_invalid
spec/models/user_spec.rb-      expect(user.errors.messages[:dob_verification]).to be_present
--
spec/models/user_spec.rb:      expect(user).to be_invalid
spec/models/user_spec.rb-      expect(user.errors.messages[:dob_verification]).to eq [I18n.t('users.errors.wca_id_no_birthdate_html')]
--
spec/models/user_spec.rb:      expect(user).to be_invalid
spec/models/user_spec.rb-      expect(user.errors.messages[:unconfirmed_wca_id]).to eq ["already assigned to a different user"]
--
spec/models/user_spec.rb:      expect(user).to be_invalid
spec/models/user_spec.rb-      expect(user.errors.messages[:dob_verification]).to be_present
--
spec/models/user_spec.rb:      expect(user).to be_invalid
spec/models/user_spec.rb-      expect(user.errors.messages[:delegate_id_to_handle_wca_id_claim]).to eq ['required']
--
spec/models/user_spec.rb:      expect(user_with_wca_id).to be_invalid
spec/models/user_spec.rb-      expect(user_with_wca_id.errors.messages[:unconfirmed_wca_id]).to eq [
--
spec/models/vote_spec.rb:    expect(vote).to be_invalid
spec/models/vote_spec.rb-    expect(vote.errors[:poll_options]).to eq ["can't be empty"]
--
spec/models/vote_spec.rb:    expect(vote).to be_invalid
spec/models/vote_spec.rb-    expect(vote.errors[:poll_options]).to eq ["you must choose just one option"]
--
spec/models/vote_spec.rb:    expect(vote).to be_invalid
spec/models/vote_spec.rb-    expect(vote.errors[:poll_id]).to eq ["poll is not confirmed"]
--
spec/models/vote_spec.rb:    expect(vote).to be_invalid
spec/models/vote_spec.rb-    expect(vote.errors[:poll_id]).to eq ["poll is closed"]
--
spec/models/vote_spec.rb:    expect(vote).to be_invalid
spec/models/vote_spec.rb-    expect(vote.errors[:poll_id]).to eq ["is not valid"]
--
spec/models/vote_spec.rb:    expect(vote).to be_invalid
spec/models/vote_spec.rb-    expect(vote.errors[:poll_options]).to eq ["One or more poll_options don't match the poll"]
```

</details>